packages/ckeditor5-list/src/todolist/todolistediting.ts

```diff
@@ -357,6 +361,35 @@ export default class TodoListEditing extends Plugin {
 			writer.setSelection( previousSelectionRanges );
 		} );
 	}
+
+	/**
+	 * Observe when user enters or leaves todo list and set proper aria value in global live announcer.
+	 * This allows screen readers to indicate when the user has entered and left the specified todo list.
+	 *
+	 * @internal
+	 */
+	private _initAriaAnnouncements( ) {
+		const { model, ui, t } = this.editor;
+		let lastFocusedCodeBlock: Element | DocumentFragment | null = null;
+
+		if ( !ui ) {
+			return;
+		}
+
+		model.document.selection.on<SelectionChangeRangeEvent>( 'change:range', () => {
+			const focusParent = model.document.selection.focus!.parent;
+			const lastElementIsTodoList = isTodoListItemElement( lastFocusedCodeBlock );
+			const currentElementIsTodoList = isTodoListItemElement( focusParent );
+
+			if ( lastElementIsTodoList && !currentElementIsTodoList ) {
+				ui.ariaLiveAnnouncer.announce( t( 'Leaving a to-do list' ) );
+			} else if ( !lastElementIsTodoList && currentElementIsTodoList ) {
+				ui.ariaLiveAnnouncer.announce( t( 'Entering a to-do list' ) );
+			}
+
+			lastFocusedCodeBlock = focusParent;
+		} );
+	}
```

Comment:
This passes unit tests but fails in real life. Please visit the todo-documentlist-simple manual test to see that the announcements are not read out.

I think this is because depending on the configuration, the list are represented in a different way in the model https://github.com/ckeditor/ckeditor5/blob/f29e0a09a1e31101fc9c58501b0cbbaeb65fd9f8/packages/ckeditor5-list/src/list/listediting.ts#L150-L159.

The utility that discovers them has to be smarter. See

https://github.com/ckeditor/ckeditor5/blob/f29e0a09a1e31101fc9c58501b0cbbaeb65fd9f8/packages/ckeditor5-list/src/todolist/todolistediting.ts#L77-L79

Reply:
Fixed